File: server/earthunt/internal/domain/stars_test.go
package domain

import "testing"

// Mirror of STAR_PARITY_CASES in apps/earthunt/lib/stars.test.ts. The client
// renders the score on the success screen and the server stores what the stats
// screen later shows, so the two implementations must agree on every row.
// When you add a case here, add the same one there.
func TestComputeStars_MatchesClient(t *testing.T) {
	cases := []struct {
		name         string
		attempts     int
		countryCount int
		hintsUsed    int
		want         int
	}{
		{"no attempts", 0, 0, 0, 3},
		{"negative attempts", -1, 5, 9, 3},
		{"perfect", 10, 10, 0, 3},
		{"90 percent no hints", 10, 9, 0, 3},
		{"90 percent one hint", 10, 9, 1, 2},
		{"70 percent no hints", 10, 7, 0, 2},
		{"70 percent two hints", 10, 7, 2, 2},
		{"70 percent three hints", 10, 7, 3, 1},
		{"60 percent", 10, 6, 0, 1},
		{"25 percent", 20, 5, 0, 1},
		// 69.56 % — truncation scored this 69 (1 star) while the client rounded
		// to 70 (2 stars). This is the case that exposed the split.
		{"rounds up across the 70 boundary", 23, 16, 0, 2},
		// 89.47 % must not round up to 90.
		{"does not round up across the 90 boundary", 19, 17, 0, 2},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := ComputeStars(tc.attempts, tc.countryCount, tc.hintsUsed)
			if got != tc.want {
				t.Errorf(
					"ComputeStars(attempts=%d, countryCount=%d, hintsUsed=%d) = %d, want %d",
					tc.attempts, tc.countryCount, tc.hintsUsed, got, tc.want,
				)
			}
		})
	}
}

func TestComputeStars_AlwaysInRange(t *testing.T) {
	for attempts := 0; attempts <= 40; attempts++ {
		for found := 0; found <= 20; found++ {
			for _, hints := range []int{0, 1, 2, 3, 10} {
				got := ComputeStars(attempts, found, hints)
				if got < 1 || got > 3 {
					t.Fatalf(
						"ComputeStars(%d, %d, %d) = %d, out of the 1..3 range",
						attempts, found, hints, got,
					)
				}
			}
		}
	}
}

File: apps/earthunt/lib/stars.test.ts
import { describe, expect, it } from "vitest"
import { getStars } from "./stars"

/**
 * Shared with server/earthunt/internal/domain/stars_test.go — the two
 * implementations must agree for every row, because the success screen renders
 * the client value while the stats screen renders what the server stored.
 * Keep the two tables identical when adding cases.
 */
export const STAR_PARITY_CASES: Array<{
  attempts: number
  countryCount: number
  hintsUsed: number
  expected: number
}> = [
  // No attempts recorded — both sides short-circuit to 3.
  { attempts: 0, countryCount: 0, hintsUsed: 0, expected: 3 },
  { attempts: -1, countryCount: 5, hintsUsed: 9, expected: 3 },

  // Perfect and near-perfect runs.
  { attempts: 10, countryCount: 10, hintsUsed: 0, expected: 3 },
  { attempts: 10, countryCount: 9, hintsUsed: 0, expected: 3 },
  // Same accuracy but a hint was used — drops to 2.
  { attempts: 10, countryCount: 9, hintsUsed: 1, expected: 2 },

  // The 70 % band.
  { attempts: 10, countryCount: 7, hintsUsed: 0, expected: 2 },
  { attempts: 10, countryCount: 7, hintsUsed: 2, expected: 2 },
  { attempts: 10, countryCount: 7, hintsUsed: 3, expected: 1 },

  // Below 70 %.
  { attempts: 10, countryCount: 6, hintsUsed: 0, expected: 1 },
  { attempts: 20, countryCount: 5, hintsUsed: 0, expected: 1 },

  // Rounding boundary, and the case that caught a real client/server split:
  // true accuracy is 69.56 %. Rounding gives 70 -> 2 stars; integer truncation
  // (what the Go server used to do) gives 69 -> 1 star, so the success screen
  // and the stats screen disagreed. Reachable in a real game — 16 countries is
  // inside the 3..20 range the level generator produces.
  { attempts: 23, countryCount: 16, hintsUsed: 0, expected: 2 },
  // Same shape one band up: 89.47 % must NOT round to 90.
  { attempts: 19, countryCount: 17, hintsUsed: 0, expected: 2 },
]

describe("getStars", () => {
  it.each(STAR_PARITY_CASES)(
    "attempts=$attempts found=$countryCount hints=$hintsUsed -> $expected stars",
    ({ attempts, countryCount, hintsUsed, expected }) => {
      expect(getStars(attempts, countryCount, hintsUsed)).toBe(expected)
    }
  )

  it("never returns a value outside 1..3", () => {
    for (let attempts = 0; attempts <= 40; attempts++) {
      for (let found = 0; found <= 20; found++) {
        for (const hints of [0, 1, 2, 3, 10]) {
          const stars = getStars(attempts, found, hints)
          expect(stars).toBeGreaterThanOrEqual(1)
          expect(stars).toBeLessThanOrEqual(3)
        }
      }
    }
  })
})
